### DISCUSSION TITLE
Schema incompatibilities after upgrade

### DISCUSSION THREAD

**Discussion by arjen1981** [prisma/prisma#12184] [Q&A] [answered]
Hello Everyone,

We have been building on top of Apollo and Prisma1 and are finally migrating to Prisma2, but are running into a lot of schema incompatibilities.

Since Prisma1 is both an ORM as a GraphQL API we were used to building upon the Prisma1 generated GraphQL schema (opencrud).

Over the years our prisma model has grown and is now used trough Apollo by two Apps and React Admin (ra-data-opencrud).

With this setup as our staring point we started to look for a way to generate a compatible GraphQL schema for Prisma2. Prisma2 has the concept of generators with a bunch already written by the community: https://www.prisma.io/docs/concepts/components/prisma-schema/generators

We started out by looking at the graphql-schema-generator, but this schema is much simpler and nowhere close to the Prisma1 generated GraphQL schema. TypeGraphQL Prisma comes closer, but there are still a lot of incompatibilities. Fixing these incompatibilities might be weeks or maybe months of work. Building our own generator would be a huge project as well.

This all made me wonder if I'm missing something? It shouldn't be this hard to migrate, should it? Not entirely sure what to do at this point:

Is there a generator out there that is compatible that I haven't found?
- Or should we just implement compatible resolvers ourselves in apollo?
- Or pick for instance TypeGraphQL and rewrite the data layer of our App?
- What is the risk of staying on Prisma1?

**Comment by nikolasburk** [comment, marked as answer]
Hey @arjen1981 👋 thanks a lot for reaching out about this!

> We have been building on top of Apollo and Prisma1 and are finally migrating to Prisma2, but are running into a lot of schema incompatibilities.

Just for context, have you been using `prisma-binding` or the Prisma 1 Client for your GraphQL API? Based on your message, it sounds like you've been using `prisma-binding` or maybe the first version of `nexus-prisma` in your project, is that correct?

> This all made me wonder if I'm missing something? It shouldn't be this hard to migrate, should it? 

Unfortunately you're not missing anything here. With Prisma now being an ORM, the responsibility of implementing the GraphQL API has entirely moved to the developer. 

Regarding the generators, these _could_ indeed be a potential solution but as you mentioned there doesn't exist one yet that would generate the same GraphQL API that was exposed by Prisma 1, so there's immediate no solution at hand.

We are working on a generator for better integration of Prisma 2+ with GraphQL Nexus, however, the development of the has been fairly slow and we still don't have a clear timeline for a first release (also the new generator won't generate any CRUD resolvers for the foreseeable future but primarily allow for the projection of Prisma models into GraphQL types). From what I've seen, the best GraphQL CRUD generators for Prisma are [Pal.js](https://paljs.com/generator/) as well as the TypeGraphQL generator that you mentioned. 

[Pothos](https://pothos-graphql.dev/) another promising GraphQL library with a [Prisma plugin](https://pothos-graphql.dev/docs/plugins/prisma) that might be worth exploring as well.

> Or should we just implement compatible resolvers ourselves in apollo? Or pick for instance TypeGraphQL and rewrite the data layer of our App?

These are both viable paths forward, which one to choose should probably depend on the specifics of your project and which path is causes less overhead.

> What is the risk of staying on Prisma1?

Prisma 1 has been in maintenance mode for a while and we've recently announced that we're going to [sunset Prisma 1 Cloud](https://github.com/prisma/prisma1/issues/5187). You can of course stay on Prisma 1, but there won't be much support for it going forward (at some point, we'll probably take it out of maintenance mode and announce its end of life meaning we won't even be doing critical bug/security fixes if they come up).

---

We understand that this situation is really frustrating and are sincerely apologizing for the inconvenience here! All that being said, we're definitely eager to support you with the migration as well as possible. I'm sharing your case with our Solutions Engineering team and will follow up later, I'm sure someone will be able to jump on a call with you to discuss your situation in more detail if you think that would be helpful.

**Reply by arjen1981** [reply]
Hello Nikolas, 

Thanks for your clear reply. We have been using the Prisma1 Client. Pal.js seem promising, we will explore this generator in more detail.

It would be appreciated if we could have a call with the Solutions Engineering team. We could share our project and code this way, which would be helpful to discuss the possible paths going forward.

I will send you a mail with our contact details. Thanks!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
